pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1418,3 +1435,248 @@ func (vs *VSphere) GetZoneByNodeName(ctx context.Context, nodeName k8stypes.Node
 func (vs *VSphere) GetZoneByProviderID(ctx context.Context, providerID string) (cloudprovider.Zone, error) {
 	return cloudprovider.Zone{}, cloudprovider.NotImplemented
 }
+
+// GetVolumeLabels returns the well known zone and region labels for given volume
+func (vs *VSphere) GetVolumeLabels(volumePath string) (map[string]string, error) {
```

Comment:
Which datastore is available in which zone/region is going to be static. Any specific reasons for computing this information every time?

Reply:
When creating a volume it is the admission controller that labels the volume with the zone information. The admission controller runs in the api-server and creates a new instance of a cloudprovider to fetch the zone information each time. So the cloudprovider cannot cache the zone information for a datastore.

A datastore is said to belong a zone if it is mounted on all ESX hosts of that zone. By that definition a datastore can belong to multiple zones. To calculate the zones of a datastore we do the following:
1. Compute the map of zone to all hosts in that zone. This is done across all ESX hosts in VC.
2. Find all the hosts to which the given datastore is mounted.
3. For each zone, if all the hosts in that zone are contained in the list of (2) - meaning the datastore is mounted on all hosts in that zone, then the zone is included.

Reply:
> creates a new instance of a cloudprovider to fetch the zone information each time. So the cloudprovider cannot cache the zone information for a datastore.

We can avoid recomputing the zone info per datastore by caching this in global variable(a map?) if you are getting new instance of a cloud provider every time. 

Another concern I have is that the GetVolumeLabels is a heavy weight operation that does multiple remote calls to VC.  If zones feature is not used, it looks like this method is invoked irrespective of whether the zones is enabled or not. Am I right? If that is the case, could we optimize it such that if zones functionality is not enabled, then skip computing volume labels?